### PR TITLE
Add Basic.return/2 and Basic.cancel_return/1

### PR DESCRIPTION
**Pull request by take-five**
This adds `:amqp_channel.register_return_handler/2` and
`:amqp_channel.unregister_return_handler/1` mirror functions
to `AMQP.Basic` module under names `Basic.return/2` and
`Basic.cancel_return/1` respectively. Similar mechanism as
in `AMQP.Basic.consume/4` is used to transform Erlang records to tuples.

### CODE REVIEW

**Comment by ono** [comment]
Thanks for the PR and sorry for the late reply as I have been away.

It looks good apart from the module implemented and function names.

It's better sitting on `Basic` module. I know that the handler API is implemented in `amqp_channel` module in Erlang client. But this `amqp` library is trying to stay with the classes and methods defined in the protocol (also influence by [Langohr](http://reference.clojurerabbitmq.info/langohr.basic.html)). The `return` method is defined under the `Basic` class: https://www.rabbitmq.com/amqp-0-9-1-quickref.html

`consume` is in `Basic` module with the same reason.

I also suggest to change function names to `return` and `cancel_return`. I am not 100% comfortable with naming `return` to a method but I think the benefits are...

- having similar naming convention to `consume` and `cancel`
- having same protocol name

Implementation itself looks good so I will merge PR if I see those changes. Ah, please leave `amqp_msg` in `core` as I think it's better to be there actually :+1:

Thank you!

(CC @pma)

File: test/channel_test.exs [hunk 1]
```diff
@@ -3,6 +3,9 @@ defmodule ChannelTest do
 
   alias AMQP.Connection
   alias AMQP.Channel
+  alias AMQP.Basic
+
+  @moduletag :capture_log
```
Comment (ono) [inline test/channel_test.exs]: why is this needed?

**Comment by take-five** [comment]
Hi @ono, thanks for the feedback.

You have a good points and I made the changes you asked for :)
Sorry for accidental `mix.lock` change. The difference comes from the fact that I'm using Elixir 1.4. I can revert `mix.lock` changes if you're worried about that.

**Comment by archSeer** [comment]
👍 

This was attempted before though, but it never got merged? something about an alternative consumer behaviour https://github.com/pma/amqp/pull/8

File: test/basic_test.exs [hunk 2]
```diff
@@ -15,4 +15,26 @@ defmodule BasicTest do
   test "basic publish to default exchange", meta do
     assert :ok = Basic.publish(meta[:chan], "", "", "ping")
   end
+
+  test "register_return_handler", meta do
```
Comment (ono) [inline test/basic_test.exs]: "basic return"

Comment (take-five) [inline test/basic_test.exs]: Fixed

**Comment by archSeer** [comment]
@take-five partly unrelated but do you have any plans or example code to
integrate this with Hare (since I saw you sumbitted a few PRs there too)?
I've been trying to get this to work without forking Hare, but Hare doesn't
expose the channel directly.

Feel free to email me so we don't spam up this issue :)

**Comment by ono** [comment]
@take-five looking good 👍  I will be merging later.

@archSeer thanks for pointing that out. i wasn't aware of the ticket - it is very helpful.

For a record, I am leaving my thoughts on https://github.com/pma/amqp/pull/8 here.

From performance point of view, the current implementation has an extra overhead on relaying a message from a handler process to a consumer process. Providing the handler behaviour (like [pma_consumer_behaviour](https://github.com/pma/amqp/commit/a4417abf1776ffb1b3008c372337e01e977b5ce2)) will solve the issue.

Beside I see most people actually implement GenServer for the consumer. It would provide conveniences to the users.

However it's used by quite big number of people already and I don't think removing `Basic.consume/3` is a good idea. If we want to add consumer GenServer to this repo, I think it should be an addition instead of a replacement.

At the same time, `return` has not been supported yet after 2 years since the discussion, I think it's a good idea to support it with the existing amqp concept now.

In terms of complexity(LOC), we might be able to make cleaner some duplicate code with macro but I prefer explicit duplication at this stage. Even if we want to refactor, I think we should do it with a separate PR.

Additionally we can try GenServer implementation in a different repo at first. Then we can discuss merging the project later. @pma is working on [wabbit](https://github.com/pma/wabbit) for the different approach for example.
